web/docs/pipelines.md

```diff
@@ -0,0 +1,216 @@
+matching operators:
+
+![Typed Pipeline](typed-pipeline.excalidraw.svg)
+
+We call any void-to-void operator sequence a **closed pipeline**. Only closed
+pipelines can execute. If a pipeline does not have a source and sink, it would
+"leak" data. If a pipeline is open, the engine auto-completes a source/sink when
+possible or rejects it.
+
+Zooming out in the above type table makes the operator types apparent:
+
+![Operator Types](operator-types.excalidraw.svg)
+
+In fact, we can define the operator type as a function of its input and output types:
+
+```
+(Input x Output) → {Source, Sink, Transformation}
+```
+
+## Multi-Schema Dataflows
+
+Tenzir dataflows are *multi-schema* in that a single pipeline can work
+with heterogeneous types of events, each of which have a different schemas.
+This allows you, for example, to perform aggregations across multiple events.
+Multi-schema dataflows require automatic schema inference at parse time. Tenzir
+parsers, such as [`json`](/formats/json) support this out of the box.
+
+This behavior is very different from execution engines that only work with
+structured data, where the unit of computation is typically a fixed set of
+tables. Schema-less systems, such as document-oriented databases, offer more
+simplicity, at the cost of performance. In the spectrum of performance and ease
+of use, Tenzir therefore [fills a gap](/why-tenzir):
+
+![Structured vs. Document-Oriented](structured-vs-document-oriented.excalidraw.svg)
+
+:::info Eclectic & Super-structured Data
+[Zed](https://amyousterhout.com/papers/zed_cidr23.pdf) has a type system similar
+to Tenzir, with the difference that Zed associates types *with every single
+value*. Unlike Zed, Tenzir uses a "data frame" abstraction and relies on
+homogeneous Arrow record batches of up to 65,535 rows.
+:::
+
+If the schema in a pipeline changes, we simply create new batch of events. The
+worst case for Tenzir is a ordered stream of schema-switching events, with every
+event having a new schema than the previous one. That said, even for those data
+streams we can efficiently build homogeneous batches when the inter-event order
+does not matter significantly. Similar to predicate pushdown, Tenzir operators
+support "orderness pushdown" to signal to upstream operators that the event
+order only matters intra-schema but not inter-schema. In this case we
+transparently demultiplex a heterogeneous stream into *N* homogeneous streams,
+each of which yields batches of up to 65k events. The `import` operator is an
+example of such an operator, and it pushes the its orderness upstream so that we
+can efficiently parse, say, a diverse stream of NDJSON records, such as
+Suricata's EVE JSON or Zeek's streaming JSON.
+
+You could call multi-schema dataflows *multiplexed* and there exist dedicated
+operators to demultiplex a stream. As of now, this is hard-coded per operator.
+For example, the [`directory`](/connectors/directory) operator demultiplexes a
```

Comment:
`directory` is a saver, not an operator. This should probably be rewritten to say `to directory`.

Reply:
I made this `to directory /tmp/dir write parquet`.